File: sdk-typescript/src/worker.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import {
  Server,
  ServerCredentials,
  Metadata,
  credentials,
  type sendUnaryData,
} from "@grpc/grpc-js";

import type { BlobCache } from "./blob-cache.js";
import {
  FlowServiceClient,
  WorkerServiceService,
  type InvokeExecuteMethodRequest,
  type InvokeExecuteMethodResponse,
  type InvokeWaitForMethodRequest,
  type InvokeWaitForMethodResponse,
  type InvokeWorkerRPCRequest,
  type InvokeWorkerRPCResponse,
  type WorkerServiceServer,
} from "./gen/dex.js";
import { workerServiceError } from "./grpc-status.js";
import { registeredAttributeIndexes, type Registry } from "./flow.js";
import type { WorkerOptions, WorkerTarget } from "./options.js";
import { ValueHydrator } from "./value-mapper.js";
import { WorkerDispatcher } from "./worker-dispatcher.js";

type WorkerState = "created" | "running" | "stopping" | "stopped" | "closed";

/**
 * Hosts registered Step and RPC handlers over the private WorkerService protocol.
 *
 * A Worker is one-shot. Call `start` for non-blocking startup or `run` to await
 * shutdown, then call `close`. Concurrent handlers must synchronize shared state.
 *
 * @example
 * ```ts
 * const worker = new Worker(registry, cache);
 * await worker.start();
 * try { console.log(worker.workerTarget.address); } finally { await worker.close(); }
 * ```
 */
export class Worker {
  /** Effective endpoint advertised to Dex. */
  public readonly workerTarget: WorkerTarget;

  private readonly server = new Server();
  private readonly flowService: InstanceType<typeof FlowServiceClient>;
  private readonly stopped: Promise<void>;
  private resolveStopped!: () => void;
  private state: WorkerState = "created";

  /**
   * Constructs a Worker without starting its listener.
   * @param registry - Flow definitions served by this Worker.
   * @param blobCache - Open cache used to hydrate large handler values.
   * @param options - Networking and startup settings; uses local defaults when omitted.
   */
  public constructor(
    public readonly registry: Registry,
    public readonly blobCache: BlobCache,
    public readonly options: WorkerOptions = {},
  ) {
    const bindAddress = options.bindAddress ?? "127.0.0.1:8803";
    this.workerTarget = options.workerTarget ?? targetFromBindAddress(bindAddress);
    this.flowService = new FlowServiceClient(
      options.serverAddress ?? "localhost:8801",
      credentials.createInsecure(),
    );
    const dispatcher = new WorkerDispatcher(
      registry,
      new ValueHydrator(this.flowService, blobCache),
    );
    this.server.addService(WorkerServiceService, workerService(dispatcher));
    this.stopped = new Promise((resolve) => {
      this.resolveStopped = resolve;
    });
  }

  /**
   * Synchronizes Attribute indexes and starts the WorkerService listener.
   * The promise resolves after successful binding; call exactly once.
   */
  public async start(): Promise<void> {
    if (this.state !== "created") {
      throw new Error(`Worker cannot start from state ${this.state}`);
    }
    const bindAddress = this.options.bindAddress ?? "127.0.0.1:8803";
    try {
      const timeoutMs = this.options.attributeIndexSyncTimeoutMs ?? 120_000;
      if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) {
        throw new RangeError("attributeIndexSyncTimeoutMs must be positive");
      }
      await syncAttributeIndexes(this.flowService, this.registry, timeoutMs);
      await bind(this.server, bindAddress);
      this.state = "running";
    } catch (failure) {
      this.state = "stopped";
      this.flowService.close();
      this.resolveStopped();
      throw new Error(`cannot start TypeScript Worker on ${bindAddress}`, { cause: failure });
    }
  }

  /** Starts the Worker and waits until `close` completes shutdown. */
  public async run(): Promise<void> {
    await this.start();
    await this.stopped;
  }

  /**
   * Gracefully stops the listener and releases the FlowService connection.
   * Calls before `start` and repeated calls after shutdown are safe.
   */
  public async close(): Promise<void> {
    if (this.state === "closed") {
      return;
    }
    if (this.state === "created") {
      this.state = "closed";
      this.flowService.close();
      this.resolveStopped();
      return;
    }
    if (this.state === "running") {
      this.state = "stopping";
      await shutdown(this.server);
      this.flowService.close();
      this.state = "stopped";
      this.resolveStopped();
    } else {
      await this.stopped;
    }
    this.state = "closed";
  }
}

function syncAttributeIndexes(
  service: InstanceType<typeof FlowServiceClient>,
  registry: Registry,
  timeoutMs: number,
): Promise<void> {
  return new Promise((resolve, reject) => {
    service.syncAttributeIndexes(
      { attributeIndexes: Object.fromEntries(registeredAttributeIndexes(registry)) },
      new Metadata(),
      { deadline: Date.now() + timeoutMs },
      (error) => {
        if (error !== null) {
          reject(error);
          return;
        }
        resolve();
      },
    );
  });
}

function workerService(dispatcher: WorkerDispatcher): WorkerServiceServer {
  return {
    invokeWaitForMethod(call, callback) {
      invoke(() => dispatcher.invokeWaitFor(call.request), callback);
    },
    invokeExecuteMethod(call, callback) {
      invoke(() => dispatcher.invokeExecute(call.request), callback);
    },
    invokeWorkerRpc(call, callback) {
      invoke(() => dispatcher.invokeRPC(call.request), callback);
    },
  };
}

function invoke<Response>(
  invocation: () => Promise<Response>,
  callback: sendUnaryData<Response>,
): void {
  invocation().then(
    (response) => callback(null, response),
    (failure: unknown) => callback(workerServiceError(failure), null),
  );
}

function bind(server: Server, address: string): Promise<number> {
  parseAddress(address);
  return new Promise((resolve, reject) => {
    server.bindAsync(address, ServerCredentials.createInsecure(), (error, port) => {
      if (error !== null) {
        reject(error);
        return;
      }
      resolve(port);
    });
  });
}

function shutdown(server: Server): Promise<void> {
  return new Promise((resolve, reject) => {
    server.tryShutdown((error) => {
      if (error !== undefined) {
        reject(error);
        return;
      }
      resolve();
    });
  });
}

function targetFromBindAddress(address: string): WorkerTarget {
  const parsed = parseAddress(address);
  const host = parsed.host === "" || parsed.host === "0.0.0.0" || parsed.host === "::"
    ? "localhost"
    : parsed.host;
  return { address: host.includes(":") ? `[${host}]:${parsed.port}` : `${host}:${parsed.port}` };
}

function parseAddress(address: string): { host: string; port: number } {
  if (address.trim() !== address || address === "") {
    throw new TypeError("Worker bind address is required without whitespace");
  }
  const bracketed = /^\[([^\]]+)]:(\d+)$/.exec(address);
  const separator = address.lastIndexOf(":");
  const host = bracketed?.[1] ?? (separator < 0 ? "" : address.slice(0, separator));
  const portText = bracketed?.[2] ?? (separator < 0 ? "" : address.slice(separator + 1));
  const port = Number(portText);
  if (!Number.isInteger(port) || port < 1 || port > 65_535) {
    throw new TypeError("Worker bind address requires port 1-65535");
  }
  return { host, port };
}

File: sdk-rust/crates/dex-sdk/src/worker.rs
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Sustainable Use License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Sustainable-Use-1.0

use std::collections::HashMap;
use std::net::SocketAddr;
use std::sync::Arc;
use std::sync::atomic::{AtomicU8, AtomicU32, Ordering};

use dex_protocol::dex::WorkerErrorResponse;
use dex_protocol::dex::flow_service_client::FlowServiceClient;
use dex_protocol::dex::worker_service_server::{WorkerService, WorkerServiceServer};
use dex_protocol::dex::{
    InvokeExecuteMethodOutput, InvokeExecuteMethodRequest, InvokeWaitForMethodOutput,
    InvokeWaitForMethodRequest, InvokeWorkerRpcRequest, InvokeWorkerRpcResponse,
    SyncAttributeIndexRequest,
};
use prost::Message;
use prost_types::Any;
use tokio::runtime::Runtime;
use tokio::sync::watch;
use tonic::transport::{Channel, Endpoint, Server};
use tonic::{Code, Request, Response, Status};

use crate::server_protocol::{negotiate_server_protocol, server_info_request_error};
use crate::value_hydrator::ValueHydrator;
use crate::worker_dispatcher::WorkerDispatcher;
use crate::worker_output::WorkerResponseStream;
use crate::{BlobCache, HandlerError, Registry, SdkError, SdkResult, WorkerOptions, WorkerTarget};

// Keep the complete worker error status below default gRPC trailer limits after
// the server wraps WorkerErrorResponse in ServiceErrorResponse.
const MAX_WORKER_ERROR_DETAIL_BYTES: usize = 1024;
const MAX_WORKER_ERROR_TYPE_BYTES: usize = 256;
const MAX_WORKER_STACK_TRACE_BYTES: usize = 4 * 1024;
const ERROR_DETAIL_TRUNCATION_MARKER: &str = "\n... error detail truncated by Dex Rust SDK ...";
const ERROR_TYPE_TRUNCATION_MARKER: &str = "\n... error type truncated by Dex Rust SDK ...";
const STACK_TRACE_TRUNCATION_MARKER: &str = "\n... stack trace truncated by Dex Rust SDK ...";

const CREATED: u8 = 0;
const RUNNING: u8 = 1;
const STOPPED: u8 = 2;

/// Hosts registered Flow Step and RPC handlers over application WorkerService.
///
/// A Worker owns an internal Tokio runtime and is one-shot. [`Self::start`] negotiates the Server
/// protocol, synchronizes Attribute indexes, then blocks while serving gRPC. Call [`Self::stop`]
/// from another thread to request graceful shutdown.
///
/// # Examples
///
/// ```no_run
/// use dex_sdk::{BlobCache, BlobCacheConfig, Registry, Worker, WorkerOptions};
/// use std::sync::Arc;
///
/// let cache = Arc::new(BlobCache::open(BlobCacheConfig::new(
///     "/tmp/dex-blobs", 64 * 1024 * 1024, 0,
/// )?)?);
/// let worker = Worker::try_new(Registry::new(), cache, WorkerOptions::new())?;
/// println!("advertise {}", worker.worker_target().address());
/// worker.start()?;
/// # Ok::<(), Box<dyn std::error::Error>>(())
/// ```
pub struct Worker {
    runtime: Runtime,
    service: RustWorkerService,
    flow_service: FlowServiceClient<Channel>,
    attribute_indexes: HashMap<String, i32>,
    attribute_index_sync_timeout: std::time::Duration,
    bind_address: SocketAddr,
    worker_target: WorkerTarget,
    shutdown: watch::Sender<bool>,
    state: AtomicU8,
    negotiated_protocol_version: AtomicU32,
}

impl Worker {
    /// Creates a Worker or panics when options or runtime initialization are invalid.
    ///
    /// # Panics
    ///
    /// Panics for an invalid bind address, zero index-sync timeout, invalid Dex endpoint, or Tokio
    /// runtime creation failure. Prefer [`Self::try_new`] for recoverable startup.
    pub fn new(registry: Registry, blob_cache: Arc<BlobCache>, options: WorkerOptions) -> Self {
        Self::try_new(registry, blob_cache, options)
            .unwrap_or_else(|error| panic!("cannot create Rust Worker: {error}"))
    }

    /// Creates a one-shot Worker without binding its listener.
    ///
    /// # Errors
    ///
    /// Returns [`SdkError::FlowDefinition`] for invalid Worker options or Service for runtime and
    /// endpoint initialization failures. The Dex connection is lazy.
    pub fn try_new(
        registry: Registry,
        blob_cache: Arc<BlobCache>,
        options: WorkerOptions,
    ) -> SdkResult<Self> {
        let runtime = Runtime::new().map_err(service_error)?;
        if options.attribute_index_sync_timeout_value().is_zero() {
            return Err(SdkError::FlowDefinition {
                message: "attribute index sync timeout must be positive".to_string(),
            });
        }
        let bind_address = options
            .bind_address_value()
            .parse::<SocketAddr>()
            .map_err(|error| SdkError::FlowDefinition {
                message: format!("invalid Worker bind address: {error}"),
            })?;
        let worker_target = options.worker_target_value().cloned().unwrap_or_else(|| {
            let host = if bind_address.ip().is_unspecified() {
                "127.0.0.1".to_string()
            } else {
                bind_address.ip().to_string()
            };
            WorkerTarget::new(format!("{host}:{}", bind_address.port()))
        });
        let endpoint = Endpoint::from_shared(endpoint_address(options.server_address_value()))
            .map_err(service_error)?;
        let flow_service = {
            let _runtime_guard = runtime.enter();
            FlowServiceClient::new(endpoint.connect_lazy())
        };
        let attribute_indexes = registry.attribute_indexes().clone();
        let hydrator = ValueHydrator::new(flow_service.clone(), blob_cache);
        let dispatcher = WorkerDispatcher::new(registry, hydrator);
        let (shutdown, _) = watch::channel(false);
        Ok(Self {
            runtime,
            service: RustWorkerService { dispatcher },
            flow_service,
            attribute_indexes,
            attribute_index_sync_timeout: options.attribute_index_sync_timeout_value(),
            bind_address,
            worker_target,
            shutdown,
            state: AtomicU8::new(CREATED),
            negotiated_protocol_version: AtomicU32::new(0),
        })
    }

    /// Returns the resolved WorkerService target that Clients should advertise to Dex.
    pub fn worker_target(&self) -> &WorkerTarget {
        &self.worker_target
    }

    /// Negotiates protocol, synchronizes Attribute indexes, and serves until stopped or failed.
    ///
    /// # Errors
    ///
    /// Returns FlowDefinition if called more than once, or Service for index synchronization,
    /// timeout, bind, and server failures.
    pub fn start(&self) -> SdkResult<()> {
        self.state
            .compare_exchange(CREATED, RUNNING, Ordering::AcqRel, Ordering::Acquire)
            .map_err(|_| SdkError::FlowDefinition {
                message: "Worker can only be started once".to_string(),
            })?;
        let mut shutdown = self.shutdown.subscribe();
        let mut flow_service = self.flow_service.clone();
        let server_info_result = self.runtime.block_on(async {
            tokio::time::timeout(
                self.attribute_index_sync_timeout,
                flow_service.get_server_info(()),
            )
            .await
        });
        let server_info = match server_info_result {
            Ok(Ok(response)) => response.into_inner(),
            Ok(Err(status)) => {
                self.state.store(STOPPED, Ordering::Release);
                return Err(service_error(server_info_request_error(status)));
            }
            Err(elapsed) => {
                self.state.store(STOPPED, Ordering::Release);
                return Err(service_error(server_info_request_error(elapsed)));
            }
        };
        let negotiated_protocol_version =
            negotiate_server_protocol(&server_info).map_err(|error| {
                self.state.store(STOPPED, Ordering::Release);
                service_error(error)
            })?;
        self.negotiated_protocol_version
            .store(negotiated_protocol_version, Ordering::Release);
        let sync_result = self.runtime.block_on(async {
            tokio::time::timeout(
                self.attribute_index_sync_timeout,
                flow_service.sync_attribute_indexes(SyncAttributeIndexRequest {
                    attribute_indexes: self.attribute_indexes.clone(),
                }),
            )
            .await
        });
        match sync_result {
            Ok(Ok(_)) => {}
            Ok(Err(status)) => {
                self.state.store(STOPPED, Ordering::Release);
                return Err(service_error(status));
            }
            Err(elapsed) => {
                self.state.store(STOPPED, Ordering::Release);
                return Err(service_error(elapsed));
            }
        }
        let result = self.runtime.block_on(
            Server::builder()
                .add_service(WorkerServiceServer::new(self.service.clone()))
                .serve_with_shutdown(self.bind_address, async move {
                    while !*shutdown.borrow() && shutdown.changed().await.is_ok() {}
                }),
        );
        self.state.store(STOPPED, Ordering::Release);
        result.map_err(service_error)
    }

    /// Requests shutdown of a running Worker and returns immediately.
    ///
    /// Calling `stop` before or after `start` is safe; [`Self::start`] observes the shutdown signal.
    pub fn stop(&self) {
        let _ = self.shutdown.send(true);
    }
}

#[derive(Clone)]
struct RustWorkerService {
    dispatcher: WorkerDispatcher,
}

#[tonic::async_trait]
impl WorkerService for RustWorkerService {
    type InvokeWaitForMethodStream = WorkerResponseStream<InvokeWaitForMethodOutput>;

    async fn invoke_wait_for_method(
        &self,
        request: Request<InvokeWaitForMethodRequest>,
    ) -> Result<Response<Self::InvokeWaitForMethodStream>, Status> {
        Ok(Response::new(
            self.dispatcher
                .invoke_wait_for(request.into_inner())
                .into_stream(worker_status),
        ))
    }

    type InvokeExecuteMethodStream = WorkerResponseStream<InvokeExecuteMethodOutput>;

    async fn invoke_execute_method(
        &self,
        request: Request<InvokeExecuteMethodRequest>,
    ) -> Result<Response<Self::InvokeExecuteMethodStream>, Status> {
        Ok(Response::new(
            self.dispatcher
                .invoke_execute(request.into_inner())
                .into_stream(worker_status),
        ))
    }

    async fn invoke_worker_rpc(
        &self,
        request: Request<InvokeWorkerRpcRequest>,
    ) -> Result<Response<InvokeWorkerRpcResponse>, Status> {
        self.dispatcher
            .invoke_rpc(request.into_inner())
            .await
            .map(Response::new)
            .map_err(worker_status)
    }
}

#[derive(Clone, PartialEq, Message)]
struct GoogleRpcStatus {
    #[prost(int32, tag = "1")]
    code: i32,
    #[prost(string, tag = "2")]
    message: String,
    #[prost(message, repeated, tag = "3")]
    details: Vec<Any>,
}

fn worker_status(error: HandlerError) -> Status {
    let message = truncate_worker_failure_field(
        &error.to_string(),
        MAX_WORKER_ERROR_DETAIL_BYTES,
        ERROR_DETAIL_TRUNCATION_MARKER,
    );
    let error_type = truncate_worker_failure_field(
        error.error_type(),
        MAX_WORKER_ERROR_TYPE_BYTES,
        ERROR_TYPE_TRUNCATION_MARKER,
    );
    let stack_trace = truncate_worker_failure_field(
        &format!("{message}\n{}", error.stack_trace()),
        MAX_WORKER_STACK_TRACE_BYTES,
        STACK_TRACE_TRUNCATION_MARKER,
    );
    let worker_error = WorkerErrorResponse {
        detail: message.clone(),
        error_type,
        stack_trace,
        retry_after_seconds: error.retry_after_seconds(),
    };
    let status = GoogleRpcStatus {
        code: Code::Unknown as i32,
        message: message.clone(),
        details: vec![Any {
            type_url: "type.googleapis.com/dex.WorkerErrorResponse".to_string(),
            value: worker_error.encode_to_vec(),
        }],
    };
    Status::with_details(Code::Unknown, message, status.encode_to_vec().into())
}

fn truncate_worker_failure_field(
    value: &str,
    maximum_bytes: usize,
    truncation_marker: &str,
) -> String {
    let encoded = value.as_bytes();
    if encoded.len() <= maximum_bytes {
        return value.to_string();
    }
    let mut prefix_length = maximum_bytes - truncation_marker.len();
    while prefix_length > 0 && encoded[prefix_length] & 0xc0 == 0x80 {
        prefix_length -= 1;
    }
    format!(
        "{}{}",
        std::str::from_utf8(&encoded[..prefix_length]).expect("UTF-8 boundary"),
        truncation_marker
    )
}

fn endpoint_address(address: &str) -> String {
    if address.contains("://") {
        address.to_string()
    } else {
        format!("http://{address}")
    }
}

fn service_error(error: impl std::fmt::Display) -> SdkError {
    SdkError::Service {
        service: crate::ServiceError::local("worker", error.to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn worker_status_bounds_all_text_fields_at_utf8_boundaries() {
        let error = HandlerError::new(
            "世界".repeat(MAX_WORKER_ERROR_TYPE_BYTES),
            "世界".repeat(MAX_WORKER_STACK_TRACE_BYTES),
        );

        let status = worker_status(error);
        let rpc_status = GoogleRpcStatus::decode(status.details()).expect("decode status");
        let worker = WorkerErrorResponse::decode(rpc_status.details[0].value.as_slice())
            .expect("decode worker error");

        assert!(worker.detail.len() <= MAX_WORKER_ERROR_DETAIL_BYTES);
        assert!(worker.error_type.len() <= MAX_WORKER_ERROR_TYPE_BYTES);
        assert!(worker.stack_trace.len() <= MAX_WORKER_STACK_TRACE_BYTES);
        assert!(
            worker
                .detail
                .ends_with("... error detail truncated by Dex Rust SDK ...")
        );
        assert!(
            worker
                .error_type
                .ends_with("... error type truncated by Dex Rust SDK ...")
        );
        assert_eq!(rpc_status.message, worker.detail);
        assert!(status.details().len() < 7 * 1024);
        assert!(!worker.detail.contains('\u{fffd}'));
        assert!(!worker.error_type.contains('\u{fffd}'));
        assert!(!worker.stack_trace.contains('\u{fffd}'));
    }

    #[test]
    fn truncation_preserves_utf8_boundaries_and_markers() {
        let truncated = truncate_worker_failure_field(
            &"世界".repeat(MAX_WORKER_STACK_TRACE_BYTES),
            MAX_WORKER_STACK_TRACE_BYTES,
            STACK_TRACE_TRUNCATION_MARKER,
        );

        assert!(truncated.len() <= MAX_WORKER_STACK_TRACE_BYTES);
        assert!(truncated.ends_with("... stack trace truncated by Dex Rust SDK ..."));
        assert!(!truncated.contains('\u{fffd}'));
    }
}
